generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/GraphSerializer.java

```diff
@@ -0,0 +1,82 @@
+package com.google.devtools.build.workspace.maven;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.common.graph.GraphBuilder;
+import com.google.common.graph.MutableGraph;
+import org.eclipse.aether.graph.DependencyNode;
+import org.eclipse.aether.graph.DependencyVisitor;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Serializes a resolved dependency graph into a set of Bazel Build Rules.
+ */
+public class GraphSerializer {
+
+  public static Set<MavenJarRule> generateBuildRules(DependencyNode root) {
+    MutableGraph<MavenJarRule> graph = GraphBuilder.directed().allowsSelfLoops(false).build();
+    DependencyVisitor visitor = new AetherGraphTraverser(graph);
+    root.getChildren().forEach(c -> c.accept(visitor));
+    return graph.nodes();
+  }
+}
+
+/**
+ * Performs a depth first traversal of aether's dependency node graph, and
+ * constructs a guava graph representation of the aether graph, with each
+ * node being a Rule object.
+ */
+class AetherGraphTraverser implements DependencyVisitor {
+  private final Set<DependencyNode> visited;
+  private final MutableGraph<MavenJarRule> ruleGraph;
+  private final Map<DependencyNode, MavenJarRule> nodeRuleMap;
+
+  AetherGraphTraverser(MutableGraph<MavenJarRule> ruleGraph) {
+    this.visited = Sets.newHashSet();
+    this.ruleGraph = ruleGraph;
+    this.nodeRuleMap = Maps.newHashMap();
+  }
+
+  @Override
+  public boolean visitEnter(DependencyNode dependencyNode) {
+    if (dependencyNode == null || dependencyNode.getDependency().isOptional()) {
+      return false;
+    }
+    boolean isFirstVisit = visited.add(dependencyNode);
+    if (!isFirstVisit) {
```

Comment:
Disappointing they never override the equals() method. 

This is a potential performance issue, since we don't want to traverse/accept a node multiple times. It doesn't affect correctness.

I suspect that object.equals() may suffice for identifying equal nodes. 

This suspicion derives from Aether's [TreeDependencyVisitor](https://github.com/sonatype/sonatype-aether/blob/master/aether-util/src/main/java/org/sonatype/aether/util/graph/TreeDependencyVisitor.java) class. 

```java
public boolean visitEnter( DependencyNode node )
    {
        boolean visited = visitedNodes.put( node, Boolean.TRUE ) != null;

        visits.push( Boolean.valueOf( visited ) );

        if ( visited )
        {
            return false;
        }

        return visitor.visitEnter( node );
    }

```
[Related tests](https://github.com/sonatype/sonatype-aether/blob/master/aether-util/src/test/java/org/sonatype/aether/util/graph/TreeDependencyVisitorTest.java)

I will include a TODO, and investigate this further.